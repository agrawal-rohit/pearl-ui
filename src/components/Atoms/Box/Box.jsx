import React from "react";
import { View } from "react-native";
import useStyledProps from "../../../hooks/useStyledProps";
import { useTheme } from "../../../hooks/useTheme";
import {
  backgroundColor,
  border,
  layout,
  opacity,
  position,
  shadow,
  spacing,
  visible,
} from "../../../theme/src/styleFunctions";
import {
  IBasePearlTheme,
  RestyleFunctionContainer,
} from "../../../theme/src/types";

// export type BoxProps<
//   Theme extends IBasePearlTheme
// > = BackgroundColorProps<Theme> &
//   OpacityProps &
//   VisibleProps &
//   LayoutProps &
//   SpacingProps<Theme> &
//   BorderProps<Theme> &
//   ShadowProps<Theme> &
//   PositionProps<Theme>;

export const boxRestyleFunctions = [
  backgroundColor,
  opacity,
  visible,
  layout,
  spacing,
  border,
  shadow,
  position,
];

/**
 * Box is the most abstract component on top of which all other Pearl UI components are built. By default, it renders a <View> element.
 */
// type ViewProps = React.ComponentProps<typeof View> & {
//   children?: React.ReactNode;
// };
// type ComponentProps<Theme extends IBasePearlTheme> = BoxProps<Theme> &
//   Omit<ViewProps, keyof BoxProps<Theme>>;

const Box = React.forwardRef((props, ref) => {
  const passedProps = useStyledProps(boxRestyleFunctions, props);

  return <View ref={ref} {...passedProps} />;
});

export default Box;
