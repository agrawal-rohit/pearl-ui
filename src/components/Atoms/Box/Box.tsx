import React from "react";
import { View, ViewProps } from "react-native";
import { pearlify } from "../../../pearlify";

/**
 * Box is the most abstract component on top of which all other Pearl UI components are built.
 * By default, it renders a <View> element.
 */
const Box = pearlify<ViewProps>(View, {
  componentName: "Box",
  type: "basic",
  animatable: true,
});

// The props that the Box component accepts.
export type BoxProps = React.ComponentProps<typeof Box>;

// Export the Box component as the default export.
export default Box;
