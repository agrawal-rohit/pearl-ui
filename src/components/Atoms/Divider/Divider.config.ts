export default {
  baseStyle: {
    bg: {
      light: "neutral.300",
      dark: "neutral.900",
    },
    thickness: 1,
    length: "100%",
  },
  defaults: {},
};
