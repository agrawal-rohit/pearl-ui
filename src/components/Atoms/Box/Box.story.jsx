import React from "react";
import { storiesOf } from "@storybook/react-native";
import Screen from "../Screen/Screen";
import Box from "./Box";

storiesOf("Box", module)
  .addDecorator((getStory) => getStory())
  .add("Basic box", () => (
    <Box
      padding="s"
      height={300}
      width={100}
      backgroundColor="primary100"
    ></Box>
  ));
