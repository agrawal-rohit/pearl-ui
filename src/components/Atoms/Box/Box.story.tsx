import React from "react";
import { storiesOf } from "@storybook/react-native";
import Screen from "../Screen/Screen";
import Box from "../Box/Box";

storiesOf("Box", module)
  .addDecorator((getStory) => <Screen>{getStory()}</Screen>)
  .add("Basic box", () => (
    <Box padding="s" width={200} backgroundColor="primary100"></Box>
  ));
