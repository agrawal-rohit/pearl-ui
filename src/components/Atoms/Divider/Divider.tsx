import React from "react";
import { useAtomicComponentConfig } from "../../../hooks/useAtomicComponentConfig";
import { View } from "react-native";
import { BoxProps } from "../../Atoms/Box/Box";

export type DividerProps = BoxProps & {
  /** The size of the divider */
  size?: string;
  /** The variant of the divider */
  variant?: string;
  /** The length of the divider */
  length?: number | string;
  /** The thickness of the divider */
  thickness?: number;
  /** The direction of the divider */
  orientation?: "horizontal" | "vertical";
};

/** Divider is used to visually separate content in a list or group. */
const Divider: React.FC<DividerProps> = ({ children, ...rest }) => {
  const props = useAtomicComponentConfig("Divider", rest, {
    size: rest.size,
    variant: rest.variant,
  });

  return (
    <View
      style={[
        props.style,
        {
          height:
            props.orientation === "horizontal" ? props.thickness : props.length,
          width:
            props.orientation === "vertical" ? props.thickness : props.length,
        },
      ]}
    >
      {children}
    </View>
  );
};

export default Divider;
