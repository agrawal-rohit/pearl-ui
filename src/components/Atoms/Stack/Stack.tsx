import React, { ReactElement } from "react";
import { FinalPearlTheme, ResponsiveValue } from "../../../theme/src/types";
import { getKeys } from "../../../theme/utils/type-helpers";
import { useStyleProps } from "../../../hooks/useStyleProps";
import Box, { BoxProps } from "../box/box";
import { positionStyleFunction } from "../../../theme/src/style-functions";

export type StackProps = BoxProps & {
  /** The direction to stack the items */
  direction: "horizontal" | "vertical";
  /** The spacing between the elements */
  spacing?: ResponsiveValue<keyof FinalPearlTheme["spacing"]>;
  /** If specified, each stack item will show the provided divider */
  divider?: React.ReactElement;
};

export type ZStackProps = BoxProps & {
  /** Specifies the stacking order of the provided elements */
  reversed?: boolean;
};

/**
 * Stack is a layout component that makes it easy to stack elements together and apply a space between them.
 */
const Stack: React.FC<StackProps> = ({ children, spacing = "2", ...rest }) => {
  const arrayChildren = React.Children.toArray(children);

  /**
   * Renders the children of the stack.
   *
   * @returns The rendered children.
   */
  const renderChildren = () => {
    return React.Children.map(arrayChildren, (child, index) => {
      const isLast = index === arrayChildren.length - 1;

      return (
        <Box
          flexDirection={rest.direction === "horizontal" ? "row" : "column"}
          mr={
            !isLast
              ? rest.direction === "horizontal"
                ? spacing
                : undefined
              : undefined
          }
          mb={
            !isLast
              ? rest.direction === "vertical"
                ? spacing
                : undefined
              : undefined
          }
        >
          {React.cloneElement(child as ReactElement)}
          {rest.divider &&
            !isLast &&
            React.cloneElement(rest.divider, {
              orientation:
                rest.direction === "horizontal" ? "vertical" : "horizontal",
              ml: rest.direction === "horizontal" ? spacing : undefined,
              mt: rest.direction === "vertical" ? spacing : undefined,
            })}
        </Box>
      );
    });
  };

  return (
    <Box
      {...rest}
      flexDirection={rest.direction === "horizontal" ? "row" : "column"}
      flexWrap={rest.direction === "horizontal" ? "wrap" : "nowrap"}
    >
      {renderChildren()}
    </Box>
  );
};

/**
 * HStack is a layout component that stacks elements horizontally and apply a space between them.
 */
export const HStack: React.FC<Omit<StackProps, "direction">> = ({
  children,
  ...rest
}) => {
  return (
    <Stack {...rest} direction="horizontal">
      {children}
    </Stack>
  );
};

/**
 * VStack is a layout component that stacks elements vertically and apply a space between them.
 */
export const VStack: React.FC<Omit<StackProps, "direction">> = ({
  children,
  ...rest
}) => {
  return (
    <Stack {...rest} direction="vertical">
      {children}
    </Stack>
  );
};

/**
 * ZStack is a layout component that stacks elements on top of each other.
 */
export const ZStack: React.FC<ZStackProps> = ({
  children,
  reversed = false,
  ...rest
}) => {
  const arrayChildren = React.Children.toArray(children);

  /**
   * Renders the children of the stack.
   *
   * @returns The rendered children.
   */
  const renderChildren = () => {
    return React.Children.map(arrayChildren, (child, index) => {
      const isOverridenZIndexProvided =
        (child as ReactElement).props &&
        getKeys((child as ReactElement).props).includes("zIndex");
      const computedZIndex = useStyleProps((child as ReactElement).props, [
        ...positionStyleFunction,
      ]);

      return React.cloneElement(child as ReactElement, {
        ...(child as ReactElement).props,
        style: {
          position: index === 0 ? "relative" : "absolute",
          zIndex: isOverridenZIndexProvided
            ? computedZIndex.style.zIndex
            : reversed
            ? arrayChildren.length - index
            : index,
          ...(child as ReactElement).props.style,
        },
      });
    });
  };

  return (
    <Box alignSelf="flex-start" {...rest}>
      {renderChildren()}
    </Box>
  );
};

export default Stack;
