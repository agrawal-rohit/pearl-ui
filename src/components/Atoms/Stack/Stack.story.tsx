import React from "react";
import { storiesOf } from "@storybook/react-native";
import Screen from "../Screen/Screen";
import Stack from "./Stack";
import Box from "../Box/Box";
import Divider from "../Divider/Divider";

storiesOf("Stack", module)
  .addDecorator((getStory) => <Screen>{getStory()}</Screen>)
  .add("Horizontal", () => (
    <Stack direction="horizontal" spacing="l" divider={<Divider />}>
      <Box w={20} h={100} backgroundColor="primary.500" />
      <Box w={20} h={20} backgroundColor="primary.500" />
      <Box w={20} h={20} backgroundColor="primary.500" />
      <Box w={20} h={20} backgroundColor="primary.500" />
    </Stack>
  ))
  .add("Vertical", () => (
    <Stack mt="l" direction="vertical" spacing="l" divider={<Divider />}>
      <Box w={100} h={20} backgroundColor="primary.500" />
      <Box w="40%" h={20} backgroundColor="primary.500" />
      <Box w={20} h={20} backgroundColor="primary.500" />
      <Box w={20} h={20} backgroundColor="primary.500" />
    </Stack>
  ));
