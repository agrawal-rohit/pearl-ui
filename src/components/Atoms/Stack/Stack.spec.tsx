import React from "react";
import { render } from "@testing-library/react-native";
import Stack, { HStack, VStack, ZStack } from "./stack";
import { ThemeProvider } from "../../../theme/src/theme-context";
import Divider from "../divider/divider";
import Box from "../box/box";

jest.useFakeTimers();

describe("Atoms/Stack", () => {
  it("passes the snapshot test for horizontal direction", () => {
    const tree = render(
      <ThemeProvider>
        <Stack direction="horizontal" spacing="l" divider={<Divider />}>
          <Box w={20} h={100} backgroundColor="primary.500" />
          <Box w={20} h={20} backgroundColor="primary.500" />
          <Box w={20} h={20} backgroundColor="primary.500" />
          <Box w={20} h={20} backgroundColor="primary.500" />
        </Stack>
      </ThemeProvider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("passes the snapshot test for vertical direction", () => {
    const tree = render(
      <ThemeProvider>
        <Stack mt="l" direction="vertical" spacing="l" divider={<Divider />}>
          <Box w={100} h={20} backgroundColor="primary.500" />
          <Box w="40%" h={20} backgroundColor="primary.500" />
          <Box w={20} h={20} backgroundColor="primary.500" />
          <Box w={20} h={20} backgroundColor="primary.500" />
        </Stack>
      </ThemeProvider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe("Atoms/HStack", () => {
  it("passes the snapshot test", () => {
    const tree = render(
      <ThemeProvider>
        <HStack spacing="l" divider={<Divider />}>
          <Box w={20} h={100} backgroundColor="primary.500" />
          <Box w={20} h={20} backgroundColor="primary.500" />
          <Box w={20} h={20} backgroundColor="primary.500" />
          <Box w={20} h={20} backgroundColor="primary.500" />
        </HStack>
      </ThemeProvider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe("Atoms/VStack", () => {
  it("passes the snapshot test", () => {
    const tree = render(
      <ThemeProvider>
        <VStack spacing="l" divider={<Divider />}>
          <Box w={20} h={100} backgroundColor="primary.500" />
          <Box w={20} h={20} backgroundColor="primary.500" />
          <Box w={20} h={20} backgroundColor="primary.500" />
          <Box w={20} h={20} backgroundColor="primary.500" />
        </VStack>
      </ThemeProvider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe("Atoms/ZStack", () => {
  it("passes the snapshot test", () => {
    const tree = render(
      <ThemeProvider>
        <ZStack>
          <Box w={100} h={100} backgroundColor="primary.500" />
          <Box w={50} h={50} backgroundColor="neutral.500" />
          <Box w={20} h={20} backgroundColor="danger.500" />
        </ZStack>

        <ZStack mt="xl" reversed>
          <Box w={100} h={100} backgroundColor="primary.500" />
          <Box w={50} h={50} backgroundColor="neutral.500" />
          <Box w={20} h={20} backgroundColor="danger.500" />
        </ZStack>

        <ZStack mt="xl">
          <Box w={100} h={100} backgroundColor="primary.500" />
          <Box w={50} h={50} zIndex="docked" backgroundColor="neutral.500" />
          <Box w={20} h={20} backgroundColor="danger.500" />
        </ZStack>
      </ThemeProvider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
