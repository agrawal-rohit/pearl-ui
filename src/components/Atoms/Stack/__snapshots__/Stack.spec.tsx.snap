// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Atoms/Stack passes the snapshot test for horizontal direction 1`] = `
<View
  direction="horizontal"
  divider={<Divider />}
  onLayout={[Function]}
  spacing="l"
  style={
    Object {
      "alignSelf": "flex-start",
      "borderRadius": undefined,
      "display": "flex",
      "flexDirection": "row",
      "flexWrap": "wrap",
    }
  }
>
  <View
    onLayout={[Function]}
    style={
      Object {
        "alignSelf": "stretch",
        "borderRadius": undefined,
        "display": "flex",
        "flexDirection": "row",
        "marginRight": 24,
      }
    }
  >
    <View
      onLayout={[Function]}
      style={
        Object {
          "backgroundColor": "#6A7BFF",
          "borderRadius": undefined,
          "display": "flex",
          "height": 100,
          "width": 20,
        }
      }
    />
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#E4E9F2",
            "display": "flex",
            "marginLeft": 24,
          },
          Object {
            "height": "100%",
            "width": 1,
          },
        ]
      }
    />
  </View>
  <View
    onLayout={[Function]}
    style={
      Object {
        "alignSelf": "stretch",
        "borderRadius": undefined,
        "display": "flex",
        "flexDirection": "row",
        "marginRight": 24,
      }
    }
  >
    <View
      onLayout={[Function]}
      style={
        Object {
          "backgroundColor": "#6A7BFF",
          "borderRadius": undefined,
          "display": "flex",
          "height": 20,
          "width": 20,
        }
      }
    />
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#E4E9F2",
            "display": "flex",
            "marginLeft": 24,
          },
          Object {
            "height": "100%",
            "width": 1,
          },
        ]
      }
    />
  </View>
  <View
    onLayout={[Function]}
    style={
      Object {
        "alignSelf": "stretch",
        "borderRadius": undefined,
        "display": "flex",
        "flexDirection": "row",
        "marginRight": 24,
      }
    }
  >
    <View
      onLayout={[Function]}
      style={
        Object {
          "backgroundColor": "#6A7BFF",
          "borderRadius": undefined,
          "display": "flex",
          "height": 20,
          "width": 20,
        }
      }
    />
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#E4E9F2",
            "display": "flex",
            "marginLeft": 24,
          },
          Object {
            "height": "100%",
            "width": 1,
          },
        ]
      }
    />
  </View>
  <View
    onLayout={[Function]}
    style={
      Object {
        "alignSelf": "stretch",
        "borderRadius": undefined,
        "display": "flex",
        "flexDirection": "row",
      }
    }
  >
    <View
      onLayout={[Function]}
      style={
        Object {
          "backgroundColor": "#6A7BFF",
          "borderRadius": undefined,
          "display": "flex",
          "height": 20,
          "width": 20,
        }
      }
    />
  </View>
</View>
`;

exports[`Atoms/Stack passes the snapshot test for vertical direction 1`] = `
<View
  direction="vertical"
  divider={<Divider />}
  onLayout={[Function]}
  spacing="l"
  style={
    Object {
      "alignSelf": "flex-start",
      "borderRadius": undefined,
      "display": "flex",
      "flexDirection": "column",
      "flexWrap": "nowrap",
      "marginTop": 24,
    }
  }
>
  <View
    onLayout={[Function]}
    style={
      Object {
        "alignSelf": "stretch",
        "borderRadius": undefined,
        "display": "flex",
        "flexDirection": "column",
        "marginBottom": 24,
      }
    }
  >
    <View
      onLayout={[Function]}
      style={
        Object {
          "backgroundColor": "#6A7BFF",
          "borderRadius": undefined,
          "display": "flex",
          "height": 20,
          "width": 100,
        }
      }
    />
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#E4E9F2",
            "display": "flex",
            "marginTop": 24,
          },
          Object {
            "height": 1,
            "width": "100%",
          },
        ]
      }
    />
  </View>
  <View
    onLayout={[Function]}
    style={
      Object {
        "alignSelf": "stretch",
        "borderRadius": undefined,
        "display": "flex",
        "flexDirection": "column",
        "marginBottom": 24,
      }
    }
  >
    <View
      onLayout={[Function]}
      style={
        Object {
          "backgroundColor": "#6A7BFF",
          "borderRadius": undefined,
          "display": "flex",
          "height": 20,
          "width": "40%",
        }
      }
    />
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#E4E9F2",
            "display": "flex",
            "marginTop": 24,
          },
          Object {
            "height": 1,
            "width": "100%",
          },
        ]
      }
    />
  </View>
  <View
    onLayout={[Function]}
    style={
      Object {
        "alignSelf": "stretch",
        "borderRadius": undefined,
        "display": "flex",
        "flexDirection": "column",
        "marginBottom": 24,
      }
    }
  >
    <View
      onLayout={[Function]}
      style={
        Object {
          "backgroundColor": "#6A7BFF",
          "borderRadius": undefined,
          "display": "flex",
          "height": 20,
          "width": 20,
        }
      }
    />
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#E4E9F2",
            "display": "flex",
            "marginTop": 24,
          },
          Object {
            "height": 1,
            "width": "100%",
          },
        ]
      }
    />
  </View>
  <View
    onLayout={[Function]}
    style={
      Object {
        "alignSelf": "stretch",
        "borderRadius": undefined,
        "display": "flex",
        "flexDirection": "column",
      }
    }
  >
    <View
      onLayout={[Function]}
      style={
        Object {
          "backgroundColor": "#6A7BFF",
          "borderRadius": undefined,
          "display": "flex",
          "height": 20,
          "width": 20,
        }
      }
    />
  </View>
</View>
`;
